docs/pallas/gpu/pipelining.md

```diff
@@ -0,0 +1,325 @@
+
+The kernel body of the warp specialized pipeline is run in parallel by all compute warpgroups, and SMEM is shared between compute warpgroups since they are scheduled within the same CUDA block.`lax.axis_index` can be used inside the kernel to obtain the warpgroup ID in order to divide up work amongst compute warpgroups.
+
+<!-- #endregion -->
+
+<!-- #region id="ZGbK5gIvFZKy" -->
+### Example: Matrix Multiplication with Warp Specialization
+
+The following example extends the previous matrix multiplication example to use warp specialization. This particular kernel uses 2 compute warpgroups, which operate on separate columns of the RHS matrix but share the same LHS. Each invocation of the pipeline therefore computes 2 adjacent blocks in the output matrix.
+
+<!-- #endregion -->
+
+<!-- #region id="NYWBqa9-bp2p" -->
+
+![matmul_ws](../../_static/pallas/gpu/pipeline_matmul_ws.svg)
+
+<!-- #endregion -->
+
+<!-- #region id="OkWmfqn7b53M" -->
+We use the `carry_coroutine` pattern to initialize the WGMMA accumulator, and copy the final accumulator from registers into SMEM. Here, the carry coroutine is defined in the function `compute_thread`. It is critical that the accumulator be created inside of the `compute_thread` function to avoid allocating it in the memory warpgroup which would waste registers. To perform the. WGMMA, we wrap the `wgmma` instruction in a `pl.run_state` in order to create an accumulator ref that is initialized to the carry value.
+
+Instead of using `pl.pallas_call` to call the kernel, we instead use the GPU-specific `plgpu.kernel` entry point. `plgpu.kernel` allows us to specify the number of warpgroups to launch per CUDA block via the `num_threads` argument, and allows us to specify a `thread_name` we can use to query the warpgroup index inside of the kernel.
+
+<!-- #endregion -->
+
+```python id="EJhWnwJlFGaT"
+def matmul_warp_specialized(a, b, tile_m=128, tile_n=128, swizzle=128,
+                            compute_wgs=2):
+  dtype = jnp.float16
+  elems_128b = swizzle // jnp.dtype(dtype).itemsize
+  tile_k = elems_128b
+  grid_m = m // tile_m
+  grid_k = k // tile_k
+  grid_n = n // tile_n
+  assert tile_m % elems_128b == 0
+
+  transforms = (
+          plgpu.TilingTransform((8, elems_128b)),
+          plgpu.SwizzleTransform(128),
+      )
+
+  def kernel(a_gmem, b_gmem, o_gmem, o_smem):
+    wg_idx = lax.axis_index("wg")
+    wg_slice = pl.ds(wg_idx * tile_n, tile_n)
+    # pl.program_id obtains the index into the pallas_call grid.
+    pid_m = pl.program_id(0)
+    pid_n = pl.program_id(1)
+
+    def compute_thread():
+      acc = plgpu.layout_cast(
+          jnp.full((tile_m, tile_n), 0, dtype=jnp.float32), plgpu.Layout.WGMMA,
+      )
+      # yield returns execution to the pipelined loop and returns
```

Comment:
nit: `yield` marks the place where the pipelined loop will be inserted. Its argument are the initial carry values, and its result is the carry value after the loop completes

Reply:
Done.